src/btree/row_key.c

```diff
@@ -270,61 +270,97 @@ __wt_row_leaf_key_work(
+             * keys: keep rolling forward.
              */
             goto next;
         }
 
         /*
-         * 4: the test for an on-page reference to a key that isn't
-         * prefix compressed.
+         * 4: the test for an on-page reference to a key that isn't prefix compressed.
          */
         if (unpack->prefix == 0) {
             /*
-             * If this is the key we originally wanted, we don't
-             * care if we're rolling forward or backward, it's
-             * what we want.  Take a copy and wrap up.
+             * If this is the key we originally wanted, we don't care if we're rolling forward or
+             * backward, it's what we want. Take a copy and wrap up.
              *
-             * If we wanted a different key, this key has a valid
-             * prefix, we can use it.
-             *	If rolling backward, take a copy of the key and
-             * switch directions, we can roll forward from this key.
-             *	If rolling forward there's a bug, we should have
-             * found this key while rolling backwards and switched
-             * directions then.
+             * If we wanted a different key, this key has a valid prefix, we can use it.
+             *
+             * If rolling backward, take a copy of the key and switch directions, we can roll
+             * forward from this key.
              *
-             * The key doesn't need to be instantiated, skip past
-             * that test.
+             * If rolling forward there's a bug, we should have found this key while rolling
+             * backwards and switched directions then.
+             *
+             * The key doesn't need to be instantiated, just return.
              */
-            WT_ERR(__wt_dsk_cell_data_ref(session, WT_PAGE_ROW_LEAF, unpack, keyb));
+            WT_RET(__wt_dsk_cell_data_ref(session, WT_PAGE_ROW_LEAF, unpack, keyb));
             if (slot_offset == 0)
-                goto done;
+                return (0);
             goto switch_and_jump;
         }
 
+        key_data = unpack->data;
+        key_size = unpack->size;
+        key_prefix = unpack->prefix;
+
+prefix_continue:
+        /*
+         * Proceed with a prefix-compressed key.
+         *
+         * Prefix compression means we don't yet have a key, but there's a special case: if the key
+         * is part of the group of compressed key prefixes we saved when reading the page into
+         * memory, we can build a key for this slot. Otherwise we have to keep rolling forward or
+         * backward.
+         */
+        slot = WT_ROW_SLOT(page, rip);
+        if (slot > page->prefix_start && slot <= page->prefix_stop) {
+            /* The row-store key can change underfoot; explicitly take a copy. */
+            copy = WT_ROW_KEY_COPY(&page->pg_row[page->prefix_start]);
+
+            __wt_row_leaf_key_info(page, copy, NULL, NULL, &group_key, &group_size, &group_prefix);
+            if (group_key != NULL) {
```

Comment:
There is no need to use `group_size` or `group_prefix`? Since we close out a group whenever it grows, should `key_prefix` and `group_prefix` be the same? Would it be worth an assertion? Also since they're not used, maybe rename them to `unused_size` and `unused_prefix`?

Reply:
I added the following comment in 0c7d88ad7:
```
            /*
             * Get the root key's information (the row-store key can change underfoot; explicitly
             * take a copy). Ignore the root key's size and prefix information because it must be
             * large enough (else the current key couldn't have been prefix-compressed based on its
             * value), and it can't have a prefix-compression value, it's a root key which is never
             * prefix-compressed.
             */
```